runtime/fiber.c

```diff
@@ -1117,3 +1133,318 @@ CAMLexport void caml_raise_unhandled_effect (value effect)
+  binding_s cache[DYNAMIC_CACHE_SIZE];
+  binding_table_s base;
+} dynamic_thread_s, *dynamic_thread_t;
+
+static void dynamic_flush_cache(dynamic_thread_t thread)
+{
+  /* Clear keys */
+  for (size_t i = 0; i < DYNAMIC_CACHE_SIZE; ++i) {
+    thread->cache[i].dyn = Val_null;
+  }
+}
+
+/* parent is NULL for the first thread when systhreads initializes */
+
+CAMLexport dynamic_thread_t caml_dynamic_new_thread(dynamic_thread_t parent)
+{
+  dynamic_thread_t res = caml_stat_alloc_noexc(sizeof(dynamic_thread_s));
+  if (!res) {
+    goto fail_alloc;
+  }
+  /* Not even going to think about the semantics of copying cache entries */
+  dynamic_flush_cache(res);
+ dynamic_table_fresh(&res->base);
+
+  if (parent && parent->base.count) { /* Existing base bindings to copy */
+    size_t capacity = parent->base.mask + 1;
+    binding_t new_bindings = caml_stat_alloc_noexc(sizeof(binding_s) * capacity);
+    if (!new_bindings) {
+      goto fail_alloc_bindings;
+    }
+    memcpy(new_bindings, parent->base.bindings, sizeof(binding_s) * capacity);
+    res->base.mask = parent->base.mask;
+    res->base.count = parent->base.count;
+    res->base.bindings = new_bindings;
+  }
+  return res;
+fail_alloc_bindings:
+  caml_stat_free(res);
+fail_alloc:
+  return NULL;
+}
+
+CAMLexport void caml_dynamic_delete_thread(dynamic_thread_t thread)
+{
+  if (thread->base.bindings) {
+    caml_stat_free(thread->base.bindings);
+  }
+  caml_stat_free(thread);
+}
+
+CAMLexport void caml_dynamic_flush_thread(dynamic_thread_t thread)
+{
+  dynamic_flush_cache(thread);
+}
+
+CAMLexport void caml_dynamic_enter_thread(dynamic_thread_t thread)
+{
+  Caml_state->dynamic_bindings = thread;
+}
+
+CAMLexport void caml_dynamic_scan_thread_roots(dynamic_thread_t thread,
+                                               scanning_action f,
+                                               scanning_action_flags fflags,
+                                               void *fdata)
+{
+  for (size_t i = 0; i < DYNAMIC_CACHE_SIZE; ++i) {
+    if (Is_bound(&thread->cache[i])) {
+      f(fdata, thread->cache[i].dyn, &thread->cache[i].dyn);
+      f(fdata, thread->cache[i].val, &thread->cache[i].val);
+    }
+  }
+  for (size_t i = 0; i < (size_t)thread->base.mask + 1; ++i) {
+    if (Is_bound(&thread->base.bindings[i])) {
+      f(fdata, thread->base.bindings[i].dyn, &thread->base.bindings[i].dyn);
+      f(fdata, thread->base.bindings[i].val, &thread->base.bindings[i].val);
+    }
+  }
+
+}
+
+CAMLprim value caml_dynamic_set_root(value dyn, value val)
+{
+  CAMLparam2(dyn, val); /* TODO: remove unless we can GC */
+  dynamic_thread_t thread = Caml_state->dynamic_bindings;
+  CAMLassert(thread);
+  bool res = dynamic_table_bind(&thread->base, dyn, val);
+  if (!res) {
+    caml_raise_out_of_memory();
+  }
+
+  /* invalidate cache */
+  uintnat hash = Hash_dyn(dyn);
+  uintnat index = hash & (DYNAMIC_CACHE_SIZE - 1);
+  binding_t entry = thread->cache + index;
+  if (entry->dyn == dyn) {
```

Comment:
Note this doesn't need to probe since the cache doesn't do chaining (or should it?)